distributed/semaphore.py

```diff
@@ -0,0 +1,327 @@
+from .utils import PeriodicCallback, log_errors, parse_timedelta
+from .worker import get_client, get_worker
+from .metrics import time
+import warnings
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class _Watch:
+    def __init__(self, duration=None):
+        self.duration = duration
+        self.started_at = None
+
+    def start(self):
+        self.started_at = time()
+
+    def leftover(self):
+        if self.duration is None:
+            return None
+        else:
+            elapsed = time() - self.started_at
+            return max(0, self.duration - elapsed)
+
+
+class SemaphoreExtension:
+    """ An extension for the scheduler to manage Semaphores
+
+    This adds the following routes to the scheduler
+
+    * semaphore_acquire
+    * semaphore_release
+    * semaphore_create
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.leases = defaultdict(deque)
+        self.events = defaultdict(asyncio.Event)
+        self.max_leases = dict()
+        self.leases_per_client = defaultdict(partial(defaultdict, deque))
+        self.scheduler.handlers.update(
+            {
+                "semaphore_create": self.create,
+                "semaphore_acquire": self.acquire,
+                "semaphore_release": self.release,
+                "semaphore_close": self.close,
+            }
+        )
+
+        self.scheduler.extensions["semaphores"] = self
+        self.pc_validate_leases = PeriodicCallback(
+            self._validate_leases,
+            1000
+            * parse_timedelta(
+                dask.config.get(
+                    "distributed.scheduler.locks.lease-validation-interval"
+                ),
+                default="s",
+            ),
+            io_loop=self.scheduler.loop,
+        )
+        self.pc_validate_leases.start()
+        self._validation_running = False
+
+    # `comm` here is required by the handler interface
+    async def create(self, comm=None, name=None, max_leases=None):
+        # We use `self.max_leases.keys()` as the point of truth to find out if a semaphore with a specific
+        # `name` has been created.
+        if name not in self.max_leases:
+            assert isinstance(max_leases, int), max_leases
+            self.max_leases[name] = max_leases
+        else:
+            if max_leases != self.max_leases[name]:
+                raise ValueError(
+                    "Inconsistent max leases: %s, expected: %s"
+                    % (max_leases, self.max_leases[name])
+                )
+
+    async def _get_lease(self, client, name, identifier):
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0.0005)  # This value is set somewhat arbitrarily
```

Comment:
What if the client get's deleted during waiting for a pending semaphore (can this happen?). In such a case it might block for the complete duration of a timeout in context of `acquire`.

Reply:
confirmed and removed